### PR TITLE
Allow passing multiple `--config` flags to daprd

### LINKED ISSUE #6091: Allow passing multiple `--config` flags to daprd
## In what area(s)?

/area runtime

## Describe the proposal

#5909 added the ability to pass multiple `--resources-path` flags to daprd, causing the sidecar to load resources (components, resiliency policies) from multiple paths.

In the same fashion, we should allow passing multiple `--config` flags to daprd, allowing it to load multiple Configuration resources (on file or through the operator). In case of conflicting options defined in the various Configuration resources, the last one wins.

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
# Description

Allows passing multiple `--config` flags to daprd, to make it load multiple Configuration resources.

- In self-hosted mode, that makes it load multiple files from disk
- In K8s mode, it requests multiple Configuration resources from the Operator
   - Users can pass comma-separated Configuration resources in the `dapr.io/configuration` annotation, e.g. `dapr.io/configuration: "myconfig1, myconfig2"`

If there are conflicts, the last object loaded always takes priority.

**Why is this PR so large?** In order for this to work in K8s mode, I had to change the Configuration type to have all properties include `omitempty` (and be nullable when needed). Otherwise, when the Operator serialized the Configuration resource into a JSON object, empty values would be included too, causing them to unnecessarily override the properties defined in Configuration resources loaded previously.

The PR includes unit tests for self-hosted, and E2E tests for K8s (added to the existing "secretstores" test)

## Issue reference

Fixes #6091

### CODE REVIEW

File: pkg/runtime/cli.go [hunk 1]
```diff
@@ -352,18 +350,21 @@ func FromFlags() (*DaprRuntime, error) {
...
+	var (
+		globalConfig *daprGlobalConfig.Configuration
+		configErr    error
+	)
+	if len(config) > 0 {
 		switch modes.DaprMode(*mode) {
 		case modes.KubernetesMode:
-			log.Debug("Loading Kubernetes config resource: " + *config)
-			globalConfig, configErr = daprGlobalConfig.LoadKubernetesConfiguration(*config, namespace, podName, operatorClient)
+			log.Debug("Loading Kubernetes config resources: " + strings.Join(config, ", "))
```
Comment (mukundansundar) [inline pkg/runtime/cli.go:363]: nitpick:
```suggestion
			log.Debug("Loading Kubernetes config resource(s): " + strings.Join(config, ", "))
```

File: pkg/runtime/cli.go [hunk 2]
```diff
@@ -352,18 +350,21 @@ func FromFlags() (*DaprRuntime, error) {
...
+			globalConfig, configErr = daprGlobalConfig.LoadKubernetesConfiguration(config, namespace, podName, operatorClient)
 		case modes.StandaloneMode:
-			log.Debug("Loading config from file: " + *config)
-			globalConfig, _, configErr = daprGlobalConfig.LoadStandaloneConfiguration(*config)
+			log.Debug("Loading config from files: " + strings.Join(config, ", "))
```
Comment (mukundansundar) [inline pkg/runtime/cli.go:366]: nitpick:
```suggestion
			log.Debug("Loading config from file(s): " + strings.Join(config, ", "))
```

File: pkg/runtime/cli.go [hunk 3]
```diff
@@ -420,7 +422,7 @@ func FromFlags() (*DaprRuntime, error) {
 	if enableAPILogging != nil {
 		runtimeConfig.EnableAPILogging = *enableAPILogging
 	} else {
-		runtimeConfig.EnableAPILogging = globalConfig.Spec.LoggingSpec.APILogging.Enabled
+		runtimeConfig.EnableAPILogging = globalConfig.Spec.LoggingSpec != nil && globalConfig.Spec.LoggingSpec.APILogging != nil && globalConfig.Spec.LoggingSpec.APILogging.Enabled
```
Comment (mukundansundar) [inline pkg/runtime/cli.go:425]: add a shorthand?

File: pkg/runtime/runtime.go [hunk 4]
```diff
@@ -393,24 +393,23 @@ func (a *DaprRuntime) setupTracing(hostAddress string, tpStore tracerProviderSto
...
 		if protocol != "http" && protocol != "grpc" {
 			return fmt.Errorf("invalid protocol %v provided for Otel endpoint", protocol)
 		}
-		isSecure := tracingSpec.Otel.IsSecure
 
 		var client otlptrace.Client
 		if protocol == "http" {
 			clientOptions := []otlptracehttp.Option{otlptracehttp.WithEndpoint(endpoint)}
-			if !isSecure {
+			if !tracingSpec.Otel.GetIsSecure() {
```
Comment (mukundansundar) [inline pkg/runtime/runtime.go:406]: is `Get` needed, just `IsSecure`?

File: pkg/config/configuration.go [hunk 5]
```diff
@@ -142,9 +143,9 @@ type APIAccessRule struct {
 }
 
 type HandlerSpec struct {
-	Name         string       `json:"name" yaml:"name"`
-	Type         string       `json:"type" yaml:"type"`
-	Version      string       `json:"version" yaml:"version"`
+	Name         string       `json:"name,omitempty" yaml:"name,omitempty"`
+	Type         string       `json:"type,omitempty" yaml:"type,omitempty"`
+	Version      string       `json:"version,omitempty" yaml:"version,omitempty"`
 	SelectorSpec SelectorSpec `json:"selector,omitempty" yaml:"selector,omitempty"`
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:149]: I need pointers when the zero value is a valid value. This is generally the case for ints (think, "number of retries" where 0 is a valid value) or booleans ("false" can be a valid value), and when it's important to know if the value is explicitly set (e.g. a boolean explicitly set to false vs a value that's not specified).

In the case of these strings, an empty value is not valid in our case (we do not accept empty component names), so we can skip using pointers, which make the code a lot more complex.

File: pkg/config/configuration_test.go [hunk 6]
```diff
@@ -58,72 +63,165 @@ func TestLoadStandaloneConfiguration(t *testing.T) {
...
+		testCases := []struct {
+			name           string
+			confFile       string
+			featureName    Feature
+			featureEnabled bool
+		}{
+			{
+				name:           "feature is enabled",
+				confFile:       "./testdata/feature_config.yaml",
+				featureName:    Feature("Actor.Reentrancy"),
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/config/configuration_test.go:141]: It is not. The tests had that flag before and since it's just used to test a parser, any string could work.

File: pkg/runtime/cli.go (hunk 1 above)
Comment (ItalyPaleAle) [reply pkg/runtime/cli.go:363]: Done

File: pkg/runtime/cli.go (hunk 2 above)
Comment (ItalyPaleAle) [reply pkg/runtime/cli.go:366]: Done

File: pkg/config/configuration.go [hunk 7]
```diff
@@ -163,75 +164,89 @@ type SelectorField struct {
...
-	Rules   []MetricsRule `json:"rules" yaml:"rules"`
+	// Defaults to true
+	Enabled *bool         `json:"enabled,omitempty" yaml:"enabled,omitempty"`
+	Rules   []MetricsRule `json:"rules,omitempty" yaml:"rules,omitempty"`
+}
+
+// GetEnabled returns true if metrics are enabled.
+func (m MetricSpec) GetEnabled() bool {
+	// Defaults to true if nil
+	return m.Enabled == nil || *m.Enabled
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:202]: Because this defaults to true if nil

File: pkg/config/configuration.go [hunk 8]
```diff
@@ -277,19 +292,15 @@ type APILoggingSpec struct {
...
-				Otel: OtelSpec{
-					IsSecure: true,
+			TracingSpec: &TracingSpec{
+				Otel: &OtelSpec{
+					IsSecure: ptr.Of(true),
 				},
 			},
-			MetricSpec: MetricSpec{
-				Enabled: true,
+			MetricSpec: &MetricSpec{
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:300]: Yes but the code later will translate `metric` to `metrics`. This seems to be the best way to ensure everything works.

File: pkg/config/configuration.go [hunk 9]
```diff
@@ -454,3 +441,115 @@ func (c Configuration) EnabledFeatures() []string {
...
+	if c.Spec.MTLSSpec == nil {
+		return MTLSSpec{}
+	}
+	return *c.Spec.MTLSSpec
+}
+
+// GetMetricsSpec returns the metrics spec.
+// It's a short-hand that includes nil-checks for safety.
+func (c Configuration) GetMetricsSpec() MetricSpec {
+	if c.Spec.MetricSpec == nil {
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:466]: `sortMetricsSpec` makes sure that `MetricSpec` (without S) is the source of truth

File: pkg/config/configuration.go [hunk 10]
```diff
@@ -267,7 +282,7 @@ type APILoggingSpec struct {
 	// When enabled, obfuscates the values of URLs in HTTP API logs, logging the route name rather than the full path being invoked, which could contain PII.
 	// Default: false.
 	// This option has no effect if API logging is disabled.
-	ObfuscateURLs bool `json:"obfuscateURLs" yaml:"obfuscateURLs"`
+	ObfuscateURLs bool `json:"obfuscateURLs,omitempty" yaml:"obfuscateURLs,omitempty"`
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:285]: Default is false here so we're good

File: pkg/runtime/cli.go (hunk 3 above)
Comment (ItalyPaleAle) [reply pkg/runtime/cli.go:425]: Added

File: pkg/acl/acl_test.go [hunk 11]
```diff
@@ -330,18 +330,23 @@ func TestParseAccessControlSpec(t *testing.T) {
 	})
 
 	t.Run("test when no access control policy has been specified", func(t *testing.T) {
-		invalidAccessControlSpec := config.AccessControlSpec{
+		accessControlList, err := ParseAccessControlSpec(nil, "http")
+		assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/acl/acl_test.go:334]: Changed

File: charts/dapr/crds/configuration.yaml [hunk 12]
```diff
@@ -375,11 +316,7 @@ spec:
                     properties:
                       endpointAddress:
                         type: string
-                    required:
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:378]: I was able to make most fields appear as required again. These are fields that are placed in arrays or dictionaries, so it's ok if they don't have "omitempty" as it shouldn't impact the functionality I was concerned with.

File: tools/codegen.mk [hunk 13]
```diff
@@ -19,7 +19,7 @@ ifeq (, $(shell which controller-gen))
 		set -e ;\
 		CONTROLLER_GEN_TMP_DIR="$$(mktemp -d)" ;\
 		cd "$$CONTROLLER_GEN_TMP_DIR" ;\
-		go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.5.0 ; \
+		go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.11.3 ; \
```
Comment (mukundansundar) [inline tools/codegen.mk:22]: will there be any issues increasing the version of controller-get?

Comment (ItalyPaleAle) [reply tools/codegen.mk:22]: Shouldn't be an issue. Some files were already compiled with a newer version https://github.com/dapr/dapr/blob/e8d10ae0ae726945b4aee26ac1baca4c86fecb7a/charts/dapr/crds/configuration.yaml#L7

Some whitespaces may be different.

File: charts/dapr/crds/configuration.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (yaron2) [reply charts/dapr/crds/configuration.yaml:378]: > This should be fine however, as we never relied on kubernetes native validation much

Literally every user and system using Dapr CRDs rely on K8s native validation 100% of the time. Please return the required fields. This has nothing to do with the struct definitions.

File: pkg/runtime/bulk_subscriber.go [hunk 14]
```diff
@@ -367,8 +367,10 @@ func (a *DaprRuntime) publishBulkMessageHTTP(ctx context.Context, bulkSubCallDat
 			sc, _ := diag.SpanContextFromW3CString(traceID)
 			var span trace.Span
 			ctx, span = diag.StartInternalCallbackSpan(ctx, "pubsub/"+psm.topic, sc, a.globalConfig.Spec.TracingSpec)
-			spans[n] = span
-			n++
+			if span != nil {
```
Comment (mukundansundar) [inline pkg/runtime/bulk_subscriber.go:370]: @ItalyPaleAle Can the bulk subscribe fix be a separate PR?

File: charts/dapr/crds/configuration.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:378]: All the fields that could be made required are there. I am afraid I can't do more than this.

Controller-gen assumes a field is not required if it has the "omitempty" tag, and anything I tried to override that didn't work.

File: pkg/config/configuration.go [hunk 15]
```diff
@@ -452,3 +439,133 @@ func (c Configuration) EnabledFeatures() []string {
...
+// It's a short-hand that includes nil-checks for safety.
+func (c Configuration) GetAPILoggingSpec() APILoggingSpec {
+	if c.Spec.LoggingSpec == nil || c.Spec.LoggingSpec.APILogging == nil {
+		return APILoggingSpec{}
+	}
+	return *c.Spec.LoggingSpec.APILogging
+}
+
+// ToYAML returns the Configuration represented as YAML.
+func (c *Configuration) ToYAML() (string, error) {
```
Comment (daixiang0) [inline pkg/config/configuration.go:498]: See a func with same name but for `ConfigurationSpec`, both are used only one time, I think no need to make it as a func.

File: pkg/config/configuration.go [hunk 16]
```diff
@@ -452,3 +439,133 @@ func (c Configuration) EnabledFeatures() []string {
...
+		return "", err
+	}
+	return string(b), nil
+}
+
+// String implements fmt.Stringer and is used for debugging. It returns the Configuration object encoded as YAML.
+func (c ConfigurationSpec) String() string {
+	enc, err := c.ToYAML()
+	if err != nil {
+		return "Failed to marshal ConfigurationSpec object to YAML: " + err.Error()
```
Comment (daixiang0) [inline pkg/config/configuration.go:568]: Please use `fmt`.

File: pkg/config/testdata/override_gen.yaml [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+apiversion: ""
+metadata:
+    name: daprsystem
+    generatename: ""
+    namespace: default
+    selflink: ""
+    uid: ""
+    resourceversion: ""
+    generation: 0
+    creationtimestamp: "0001-01-01T00:00:00Z"
```
Comment (daixiang0) [inline pkg/config/testdata/override_gen.yaml:11]: I see some gen files cover timestamp, some not, do we need it?

Comment (ItalyPaleAle) [reply pkg/config/testdata/override_gen.yaml:11]: I don't remember how I created this file, but if it includes all this `metadata` is probably because I dumped it from the Go struct.

It should be irrelevant - the tests do not care about the `metadata` and just look at the `spec`.

File: pkg/config/configuration.go (hunk 16 above)
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:568]: Done

File: pkg/config/configuration.go (hunk 15 above)
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:498]: They are primarily used for debugging and testing.

File: charts/dapr/crds/configuration.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (yaron2) [reply charts/dapr/crds/configuration.yaml:394]: Ok, to be clear - if there are any fields in the spec that were required and enforced before this change and are now not required and were made optional, that's a breaking change to the spec and a regression. If that's not the case, we can go forward with this change.

[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:394]: Making a field that was optional, required is a breaking change, because things that worked before may not anymore. Changing fields from required to optional is not a breaking change, because what worked before will continue to work.

In this case, I don't think there's really any net effect, because before we were not requiring these fields anyways.

A good example is that the field `mtls.enabled` used to be required and now it's not anymore: https://github.com/dapr/dapr/pull/6101/files#diff-98b7452caa61c9b7f101305e14dc06249d70a8a2fe93b806857e1317ac1a68bdL314-L324 However, `mtls` itself was not required (or users would have had to specify a `mtls` field in every Configuration resource!): what that meant is that before, if users wrote `mtls` then they **had** to also type `mtls.enabled` explicitly. This should not cause any issue to users, and could actually make things easier if users try to combine multiple Configuration resources in which one enables mTLS and the other one sets the clock skew, for example.

All other fields that have the "required" tag removed are in the same boat.

[the remark this answers is not in the document]
Comment (yaron2) [reply charts/dapr/crds/configuration.yaml:394]: >  Changing fields from required to optional is not a breaking change, because what worked before will continue to work.

Disagree, in that case invalid YAML definitions that would lead to a misconfigured state that would have been previously caught at the API level now become runtime level errors, so that's a definite regression to user experience that we cannot allow.

> because before we were not requiring these fields anyways.

This is what matters - if these fields were marked as required but not treated as such before, then this change is ok. I'll use the modified CRDs from this PR and test it out to validate.

[the remark this answers is not in the document]
Comment (yaron2) [reply charts/dapr/crds/configuration.yaml:394]: When testing the modified CRDs it's clearly observed that there are breaking changes to the specs of all CRDs. As one example, with the modified CRDs it is possible to deploy the following YAML which omits the `endpointAddress` field, whereas in 1.11 this would result in an error.

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: tracing
spec:
  tracing:
    samplingRate: "1"
    otel:
      isSecure: false
      protocol: grpc
```

[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:394]: > with the modified CRDs it is possible to deploy the following YAML which omits the endpointAddress field, whereas in 1.11 this would result in an error.

How is that a breaking change though? If it were the other way around (before it could be deployed, now it can't), it would be, but this is making something more lax.

In this case, otel would not be enabled because the endpoint is missing. We have never relied on K8s validation alone because it doesn't work in standalone mode.

[the remark this answers is not in the document]
Comment (yaron2) [reply charts/dapr/crds/configuration.yaml:394]: It's a breaking change to the CRD spec as it can and will result in any type of validation to now allow misconfigurations into the cluster state where it would once break.

To focus on the actual issue and avoid discussion over semantics, even if for the sake of the argument we say there's no breaking change here, this is just as extremely unacceptable as it results in a degraded user experience that leads users down the non-happy path. Kubernetes users rely on them to prevent misconfiguration using native tooling. Required fields are an extremely common thing in schemas with a clear purpose of explicitly describing what needs to be filled to arrive at a working desired state.

[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:394]: How would you recommend we make this thing work then?

**Comment by ItalyPaleAle** [comment]
Due to feedback, closing this PR in favor of #6607 . That's a slimmed-down version of this PR which drops support for Kubernetes mode (only works in standalone mode), to avoid modifying the CRDs
